Demos/libraries/setup.py

```diff
@@ -1,33 +1,60 @@
-from __future__ import absolute_import, print_function
+# Run as:
+#    python setup.py build_ext --inplace
 
 import os
 import sys
+import sysconfig
 
 from distutils.core import setup
 from distutils.extension import Extension
 from Cython.Build import cythonize
 
 
 # For demo purposes, we build our own tiny library.
+libname_no_ext = "mymath"
+libname = f"lib{libname_no_ext}.a"  # gcc is default
+# referebce Python language native platform Python.h header include directory
+py_include = sysconfig.get_path("include")
+c_sources = f"mymath.c"  # space separated list of source files
 try:
-    print("building libmymath.a")
-    assert os.system("gcc -shared -fPIC -c mymath.c -o mymath.o") == 0
-    assert os.system("ar rcs libmymath.a mymath.o") == 0
-except:
-    if not os.path.exists("libmymath.a"):
-        print("Error building external library, please create libmymath.a manually.")
-        sys.exit(1)
+    if not sys.platform.startswith("win"):
+        print(f"building {libname}")
+        assert (
+            os.system(
+                f"gcc -I\"{"\" -I\"".join(py_include.split(os.path.pathsep))}\" -shared -fPIC -c {c_sources} -o mymath.o"
+            )
+            == 0
+        )
```

Comment:
This is barely readable to me – way too many special characters and escapes. And it hides an important call (with side effects) inside of an `assert` statement, which may not get executed in some configurations. (I'm aware that it was like this before, but now is the best time to clean it up.) We should also use `subprocess` instead of `os.system`.

```suggestion
            c_includes = [f'-I{path}' for path in py_include.split(os.path.pathsep)]
            subprocess.check_call(["gcc", "-shared", "-fPIC", "-c"] + c_includes + [c_sources, "-o", "mymath.o"])
```

By its name, `c_sources` should rather be a list of strings than a string. That would also prevent (hypothetical) issues with spaces in file paths etc.

The same applies to the main `Demos/setup.py` script.

Reply:
We can't assume that on Windows include directories will not have spaces, hence, need to wrap in quotes.
So, some hairiness may be a bit of a tax on the eyesight, but does help to ensure error free execution.
Thanks for the suggestion regarding subprocess. I will take a look if I can do a jump to setuptools at once, or use this suggestion, since it does look more like a change, over iterative fix.

Reply:
> We can't assume that on Windows include directories will not have spaces, hence, need to wrap in quotes.

I don't think we need to do that if we pass the arguments separately (as in the `subprocess` call that I suggested), rather than pushing them through a command line string.